esp-metadata/devices/esp32c2.toml

```diff
@@ -89,7 +89,73 @@ memory_map = { ranges = [
+clocks = { system_clocks = { clock_tree = [
+    # High-speed clock sources
+    { name = "XTL_CLK",     type = "source",  values = "26, 40", output = "VALUE * 1_000_000", always_on = true },
+    { name = "PLL_CLK",     type = "derived", from = "XTL_CLK",  output = "480_000_000" },
+    { name = "RC_FAST_CLK", type = "source",                     output = "17_500_000" },
+
+    # Low-speed clocks
+    { name = "OSC_SLOW_CLK",    type = "source",  output = "32768" }, # external slow clock from GPIO0
+    { name = "RC_SLOW_CLK",     type = "source",  output = "136_000" }, # adjustable via RTC_CNTL_ANA_CLK_DIV
+    { name = "RC_FAST_DIV_CLK", type = "divider", output = "RC_FAST_CLK / 256" }, # 256 is just the default value? RTC_CNTL_FOSC_DIV
+
+    # CPU clock source dividers
+    { name = "SYSTEM_PRE_DIV_IN", type = "mux", variants = [
+        { name = "XTAL",    outputs = "XTL_CLK" },
+        { name = "RC_FAST", outputs = "RC_FAST_CLK" },
+    ] },
+    { name = "SYSTEM_PRE_DIV", type = "divider", divisors = "0 .. 1024", output = "SYSTEM_PRE_DIV_IN / (DIVISOR + 1)" },
+    { name = "CPU_PLL_DIV",    type = "divider", divisors = "4, 6", output = "PLL_CLK / DIVISOR" },
+
+    # The meta-switch
+    { name = "CPU_CLK", type = "mux", always_on = true, variants = [
+        # source clock      CPU clock                   additional config (mux = variant name, divider = divisor expression)
+        { name = "XTAL",    outputs = "SYSTEM_PRE_DIV", configures = [ "APB_CLK = CPU",     "CRYPTO_CLK = CPU",     "MSPI_CLK = CPU", "SYSTEM_PRE_DIV_IN = XTAL"    ] },
+        { name = "RC_FAST", outputs = "SYSTEM_PRE_DIV", configures = [ "APB_CLK = CPU",     "CRYPTO_CLK = CPU",     "MSPI_CLK = CPU", "SYSTEM_PRE_DIV_IN = RC_FAST" ] },
+        { name = "PLL",     outputs = "CPU_PLL_DIV",    configures = [ "APB_CLK = PLL_40M", "CRYPTO_CLK = PLL_80M", "MSPI_CLK = CPU_DIV2"                           ] },
+    ] },
+
+    # CPU-dependent clock signals
+    { name = "PLL_40M",  type = "derived", from = "CPU_CLK", output = "40_000_000" }, # PLL / 12
+    { name = "PLL_60M",  type = "derived", from = "CPU_CLK", output = "60_000_000" }, # PLL / 8
+    { name = "PLL_80M",  type = "derived", from = "CPU_CLK", output = "80_000_000" }, # PLL / 6
+    { name = "CPU_DIV2", type = "divider", output = "CPU_CLK / 2" },
+
+    { name = "APB_CLK", type = "mux", variants = [
+        { name = "PLL_40M", outputs = "PLL_40M" },
+        { name = "CPU",     outputs = "CPU_CLK" },
+    ] },
+    { name = "CRYPTO_CLK", type = "mux", variants = [
+        { name = "PLL_80M", outputs = "PLL_80M" },
+        { name = "CPU",     outputs = "CPU_CLK" },
+    ] },
+    { name = "MSPI_CLK", type = "mux", variants = [
+        { name = "CPU_DIV2", outputs = "CPU_DIV2" },
+        { name = "CPU",      outputs = "CPU_CLK" },
+    ] },
+    
+    # Low-power clocks
+    { name = "RC_FAST_CLK_DIV_N", type = "divider", divisors = "0 ..= 3", output = "RC_FAST_CLK / (DIVISOR + 1)" }, # RC_DIV
```

Comment:
1: Should this generate a 4-element enum? We could set a limit of 8 or so to force an enum instead of a struct, and enums still have the `fn new(from_number: u32)` constructor.
2: Should this be `1 ..= 4`? There are arguments for keeping close to the register value, but also for the more human consumable effective divider value...

Reply:
Didn't know about this plan - interesting. But the user facing API can and will be refined later